src/Simple.OData.Client.V4.Adapter/CommandFormatter.cs

```diff
@@ -121,38 +128,82 @@ protected override void FormatInlineCount(IList<string> commandClauses)
         }
 
         private IList<string> SelectPathSegmentColumns(
-            IList<string> columns, string path, IList<string> excludePaths = null)
+            IList<string> columns, string path, EntityCollection collection, IList<string> excludePaths = null)
         {
             if (string.IsNullOrEmpty(path))
             {
                 var resultColumns = columns.Where(x => !HasMultipleSegments(x)).ToList();
                 if (excludePaths != null)
-                    resultColumns.AddRange(columns.Where(x => HasMultipleSegments(x) &&
-                        !excludePaths.Any(y => FormatFirstSegment(y).Contains(FormatFirstSegment(x)))));
+                {
+                    if (excludePaths.Count == 1 && excludePaths[0] == StarString)
+                    {
+                        var firstSegmentPropertyNames = _session.Metadata.GetNavigationPropertyNames(collection.Name).Select(x => FormatFirstSegment(x));
```

Comment:
Hosting this LINQ expression into a local variable does not convert `Select`'s deferred execution into an immediate one. `Select` will still be called--in the worst case--`columns.Count` by `firstSegmentPropertyNames.Count()` times.

Using an `ISet` would ensure that `_session.Metadata.GetNavigationPropertyNames` was only iterated once.
```csharp
var firstSegmentPropertyNames = new HashSet<string>(...);
```

Reply:
Done